File: client/go/NatsUpstreamTransport.go
package modcdp

import (
	"context"
	"crypto/tls"
	"encoding/json"
	"fmt"
	"net"
	"net/url"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/gobwas/ws"
	"github.com/gobwas/ws/wsutil"
)

const DefaultNATSURL = "ws://127.0.0.1:4223"
const DefaultNATSSubjectPrefix = "modcdp.default"
const DefaultNATSWaitTimeoutMS = 10_000

type NatsUpstreamTransport struct {
	UpstreamTransport
	URL           string
	SubjectPrefix string
	Role          string
	WaitTimeoutMS int
	Conn          net.Conn
	IsWebSocket   bool
	buffer        string
	connected     bool
	closed        bool
	writeMu       sync.Mutex
	peerCh        chan struct{}
	peerOnce      sync.Once
}

type NatsUpstreamTransportOptions struct {
	URL           string
	SubjectPrefix string
	Role          string
	WaitTimeoutMS int
}

func NewNatsUpstreamTransport(options NatsUpstreamTransportOptions) *NatsUpstreamTransport {
	normalizedURL, subjectPrefix := normalizeNATSURL(firstNonEmptyString(options.URL, DefaultNATSURL), options.SubjectPrefix)
	role := firstNonEmptyString(options.Role, "client")
	waitTimeoutMS := options.WaitTimeoutMS
	if waitTimeoutMS == 0 {
		waitTimeoutMS = DefaultNATSWaitTimeoutMS
	}
	return &NatsUpstreamTransport{
		URL:           normalizedURL,
		SubjectPrefix: subjectPrefix,
		Role:          role,
		WaitTimeoutMS: waitTimeoutMS,
		peerCh:        make(chan struct{}),
	}
}

func (t *NatsUpstreamTransport) Update(config map[string]any) {
	if config == nil {
		return
	}
	natsURL, _ := config["nats_url"].(string)
	subjectPrefix, _ := config["nats_subject_prefix"].(string)
	if natsURL != "" || subjectPrefix != "" {
		t.URL, t.SubjectPrefix = normalizeNATSURL(firstNonEmptyString(natsURL, t.URL), firstNonEmptyString(subjectPrefix, t.SubjectPrefix))
	}
}

func (t *NatsUpstreamTransport) GetInjectorConfig() ExtensionInjectorConfig {
	return ExtensionInjectorConfig{NATSURL: t.URL, NATSSubjectPrefix: t.SubjectPrefix}
}

func (t *NatsUpstreamTransport) Connect() error {
	if t.connected {
		return nil
	}
	if !validNATSSubjectPrefix(t.SubjectPrefix) {
		return fmt.Errorf("invalid NATS subject prefix %q", t.SubjectPrefix)
	}
	parsed, err := url.Parse(t.URL)
	if err != nil {
		return err
	}
	switch parsed.Scheme {
	case "ws", "wss":
		conn, _, _, err := ws.Dial(context.Background(), t.URL)
		if err != nil {
			return err
		}
		t.Conn = conn
		t.IsWebSocket = true
		if err := t.writeProtocol("CONNECT " + mustJSON(connectNATSOptions()) + "\r\nPING\r\n"); err != nil {
			_ = conn.Close()
			return err
		}
		go t.readWebSocketLoop()
	case "nats", "tls":
		host := parsed.Hostname()
		if host == "" {
			host = "127.0.0.1"
		}
		port := parsed.Port()
		if port == "" {
			port = "4222"
		}
		var conn net.Conn
		if parsed.Scheme == "tls" {
			conn, err = tls.Dial("tcp", net.JoinHostPort(host, port), &tls.Config{ServerName: host})
		} else {
			conn, err = net.Dial("tcp", net.JoinHostPort(host, port))
		}
		if err != nil {
			return err
		}
		t.Conn = conn
		if err := t.writeProtocol("CONNECT " + mustJSON(connectNATSOptions()) + "\r\nPING\r\n"); err != nil {
			_ = conn.Close()
			return err
		}
		go t.readTCPLoop()
	default:
		return fmt.Errorf("upstream.mode=nats requires ws://, wss://, nats://, or tls:// URL, got %s", t.URL)
	}
	t.connected = true
	if err := t.subscribe(); err != nil {
		return err
	}
	return t.publish(t.outgoingSubject(), map[string]any{"type": "modcdp.nats.hello", "role": t.Role, "version": 1})
}

func (t *NatsUpstreamTransport) Send(message map[string]any) error {
	if !t.connected || t.Conn == nil {
		return fmt.Errorf("NATS transport is not connected")
	}
	return t.publish(t.outgoingSubject(), map[string]any{"type": "modcdp.nats.message", "message": message})
}

func (t *NatsUpstreamTransport) WaitForPeer() error {
	select {
	case <-t.peerCh:
		return nil
	case <-time.After(time.Duration(t.WaitTimeoutMS) * time.Millisecond):
		return fmt.Errorf("timed out waiting %dms for NATS ModCDP peer", t.WaitTimeoutMS)
	}
}

func (t *NatsUpstreamTransport) Close() error {
	t.closed = true
	t.connected = false
	if t.Conn != nil {
		_ = t.Conn.Close()
		t.Conn = nil
	}
	return nil
}

func (t *NatsUpstreamTransport) subscribe() error {
	return t.writeProtocol("SUB " + t.incomingSubject() + " 1\r\n")
}

func (t *NatsUpstreamTransport) publish(subject string, message map[string]any) error {
	body, err := json.Marshal(message)
	if err != nil {
		return err
	}
	return t.writeProtocol(fmt.Sprintf("PUB %s %d\r\n%s\r\n", subject, len(body), string(body)))
}

func (t *NatsUpstreamTransport) writeProtocol(data string) error {
	if t.Conn == nil {
		return fmt.Errorf("NATS transport is not connected")
	}
	t.writeMu.Lock()
	defer t.writeMu.Unlock()
	if t.IsWebSocket {
		return wsutil.WriteClientText(t.Conn, []byte(data))
	}
	_, err := t.Conn.Write([]byte(data))
	return err
}

func (t *NatsUpstreamTransport) incomingSubject() string {
	if t.Role == "client" {
		return t.SubjectPrefix + ".browser_to_client"
	}
	return t.SubjectPrefix + ".client_to_browser"
}

func (t *NatsUpstreamTransport) outgoingSubject() string {
	if t.Role == "client" {
		return t.SubjectPrefix + ".client_to_browser"
	}
	return t.SubjectPrefix + ".browser_to_client"
}

func (t *NatsUpstreamTransport) readWebSocketLoop() {
	for !t.closed {
		data, _, err := wsutil.ReadServerData(t.Conn)
		if err != nil {
			if !t.closed {
				t.emitClose(err)
			}
			return
		}
		t.buffer = t.consumeProtocol(t.buffer + string(data))
	}
}

func (t *NatsUpstreamTransport) readTCPLoop() {
	chunk := make([]byte, 65536)
	for !t.closed {
		n, err := t.Conn.Read(chunk)
		if err != nil {
			if !t.closed {
				t.emitClose(err)
			}
			return
		}
		t.buffer = t.consumeProtocol(t.buffer + string(chunk[:n]))
	}
}

func (t *NatsUpstreamTransport) consumeProtocol(buffer string) string {
	for {
		lineEnd := strings.Index(buffer, "\r\n")
		if lineEnd < 0 {
			return buffer
		}
		line := buffer[:lineEnd]
		upper := strings.ToUpper(line)
		if strings.HasPrefix(upper, "MSG ") {
			parts := strings.Fields(line)
			size, err := strconv.Atoi(parts[len(parts)-1])
			payloadStart := lineEnd + 2
			payloadEnd := payloadStart + size
			if err != nil || len(buffer) < payloadEnd+2 {
				return buffer
			}
			payload := buffer[payloadStart:payloadEnd]
			buffer = buffer[payloadEnd+2:]
			t.handlePayload(payload)
			continue
		}
		buffer = buffer[lineEnd+2:]
		if upper == "PING" {
			_ = t.writeProtocol("PONG\r\n")
		} else if strings.HasPrefix(upper, "-ERR") {
			t.emitClose(fmt.Errorf("NATS error: %s", line))
		}
	}
}

func (t *NatsUpstreamTransport) handlePayload(payload string) {
	var parsed any
	if err := json.Unmarshal([]byte(payload), &parsed); err != nil {
		return
	}
	record, _ := parsed.(map[string]any)
	if record["type"] == "modcdp.nats.hello" {
		t.peerOnce.Do(func() { close(t.peerCh) })
		return
	}
	message := parsed
	if record["type"] == "modcdp.nats.message" {
		message = record["message"]
	}
	if cdpMessage, ok := message.(map[string]any); ok {
		t.emitRecv(cdpMessage)
	}
}

func connectNATSOptions() map[string]any {
	return map[string]any{"verbose": false, "pedantic": false, "lang": "modcdp", "version": "1", "protocol": 1}
}

func normalizeNATSURL(rawURL string, subjectPrefix string) (string, string) {
	parsed, err := url.Parse(rawURL)
	if err != nil {
		return rawURL, sanitizeNATSSubjectPrefix(firstNonEmptyString(subjectPrefix, DefaultNATSSubjectPrefix))
	}
	query := parsed.Query()
	subject := firstNonEmptyString(subjectPrefix, query.Get("subject"), query.Get("subject_prefix"), DefaultNATSSubjectPrefix)
	query.Del("subject")
	query.Del("subject_prefix")
	parsed.RawQuery = query.Encode()
	if parsed.Path == "" && (parsed.Scheme == "ws" || parsed.Scheme == "wss") {
		parsed.Path = "/"
	}
	return parsed.String(), sanitizeNATSSubjectPrefix(subject)
}

func sanitizeNATSSubjectPrefix(value string) string {
	return strings.TrimSpace(value)
}

func validNATSSubjectPrefix(value string) bool {
	subject := strings.TrimSpace(value)
	return subject != "" && !strings.ContainsAny(subject, " \t\r\n*>")
}

func mustJSON(value any) string {
	body, _ := json.Marshal(value)
	return string(body)
}

File: client/python/modcdp/NatsUpstreamTransport.py
from __future__ import annotations

import json
import socket
import ssl
import threading
from collections.abc import Mapping
from typing import Any, cast
from urllib.parse import parse_qs, urlencode, urlparse, urlunparse

from websocket import WebSocket

from .UpstreamTransport import UpstreamTransport


DEFAULT_NATS_URL = "ws://127.0.0.1:4223"
DEFAULT_NATS_SUBJECT_PREFIX = "modcdp.default"
DEFAULT_NATS_WAIT_TIMEOUT_MS = 10_000


class NatsUpstreamTransport(UpstreamTransport):
    mode = "nats"
    endpoint_kind = "modcdp_server"

    def __init__(
        self,
        options: str | Mapping[str, Any] | None = None,
    ) -> None:
        super().__init__()
        normalized_options = {"url": options} if isinstance(options, str) else dict(options or {})
        normalized_url, normalized_subject_prefix = _normalize_nats_url(
            cast(str | None, normalized_options.get("url")) or DEFAULT_NATS_URL,
            cast(str | None, normalized_options.get("subject_prefix")),
        )
        self.url = normalized_url
        self.subject_prefix = normalized_subject_prefix
        self.role = str(normalized_options.get("role") or "client")
        self.wait_timeout_ms = int(normalized_options.get("wait_timeout_ms") or DEFAULT_NATS_WAIT_TIMEOUT_MS)
        self.socket: WebSocket | socket.socket | None = None
        self.connected = False
        self.closed = False
        self.peer_seen = threading.Event()
        self.write_lock = threading.Lock()
        self.buffer = ""

    def update(self, config: dict[str, Any] | None = None) -> "NatsUpstreamTransport":
        config = config or {}
        nats_url = config.get("nats_url")
        nats_subject_prefix = config.get("nats_subject_prefix")
        if isinstance(nats_url, str) and nats_url or isinstance(nats_subject_prefix, str) and nats_subject_prefix:
            current_url = self.url or DEFAULT_NATS_URL
            self.url, self.subject_prefix = _normalize_nats_url(
                nats_url if isinstance(nats_url, str) and nats_url else current_url,
                nats_subject_prefix if isinstance(nats_subject_prefix, str) and nats_subject_prefix else self.subject_prefix,
            )
        return self

    def getInjectorConfig(self) -> dict[str, Any]:
        return {"nats_url": self.url, "nats_subject_prefix": self.subject_prefix}

    def connect(self) -> None:
        if self.connected:
            return
        parsed = urlparse(self.url)
        if parsed.scheme in ("ws", "wss"):
            ws = WebSocket()
            ws.connect(self.url)
            self.socket = ws
            self._write_protocol(f"CONNECT {json.dumps(_connect_options())}\r\nPING\r\n")
            threading.Thread(target=self._read_websocket_loop, daemon=True).start()
        elif parsed.scheme in ("nats", "tls"):
            port = parsed.port or 4222
            host = cast(str, parsed.hostname or "127.0.0.1")
            raw_socket = socket.create_connection((host, port))
            tcp_socket = ssl.create_default_context().wrap_socket(raw_socket, server_hostname=host) if parsed.scheme == "tls" else raw_socket
            self.socket = tcp_socket
            self._write_protocol(f"CONNECT {json.dumps(_connect_options())}\r\nPING\r\n")
            threading.Thread(target=self._read_tcp_loop, daemon=True).start()
        else:
            raise RuntimeError(f"upstream.mode=nats requires ws://, wss://, nats://, or tls:// URL, got {self.url}.")
        self.connected = True
        self._subscribe()
        self._publish(self._outgoing_subject(), {"type": "modcdp.nats.hello", "role": self.role, "version": 1})

    def send(self, message: dict[str, Any]) -> None:
        if not self.connected or self.socket is None:
            raise RuntimeError("NATS transport is not connected.")
        self._publish(self._outgoing_subject(), {"type": "modcdp.nats.message", "message": message})

    def waitForPeer(self) -> None:
        if not self.peer_seen.wait(self.wait_timeout_ms / 1000):
            raise RuntimeError(f"Timed out waiting {self.wait_timeout_ms}ms for NATS ModCDP peer.")

    def close(self) -> None:
        self.closed = True
        try:
            if isinstance(self.socket, WebSocket):
                self.socket.close()
            elif self.socket is not None:
                self.socket.close()
        except Exception:
            pass
        self.socket = None
        self.connected = False
        self.peer_seen.clear()

    def _subscribe(self) -> None:
        self._write_protocol(f"SUB {self._incoming_subject()} 1\r\n")

    def _publish(self, subject: str, message: dict[str, Any]) -> None:
        body = json.dumps(message, separators=(",", ":"))
        self._write_protocol(f"PUB {subject} {len(body.encode())}\r\n{body}\r\n")

    def _write_protocol(self, data: str) -> None:
        if self.socket is None:
            raise RuntimeError("NATS transport is not connected.")
        with self.write_lock:
            if isinstance(self.socket, WebSocket):
                self.socket.send(data)
            else:
                self.socket.sendall(data.encode())

    def _incoming_subject(self) -> str:
        return f"{self.subject_prefix}.{'browser_to_client' if self.role == 'client' else 'client_to_browser'}"

    def _outgoing_subject(self) -> str:
        return f"{self.subject_prefix}.{'client_to_browser' if self.role == 'client' else 'browser_to_client'}"

    def _read_websocket_loop(self) -> None:
        try:
            while not self.closed and isinstance(self.socket, WebSocket):
                data = self.socket.recv()
                if isinstance(data, bytes):
                    self.buffer += data.decode()
                else:
                    self.buffer += str(data)
                self.buffer = self._consume_protocol(self.buffer)
        except Exception as error:
            if not self.closed:
                self._emit_close(error if isinstance(error, Exception) else RuntimeError(str(error)))

    def _read_tcp_loop(self) -> None:
        try:
            while not self.closed and isinstance(self.socket, socket.socket):
                chunk = self.socket.recv(65536)
                if not chunk:
                    break
                self.buffer += chunk.decode()
                self.buffer = self._consume_protocol(self.buffer)
        except Exception as error:
            if not self.closed:
                self._emit_close(error if isinstance(error, Exception) else RuntimeError(str(error)))

    def _consume_protocol(self, buffer: str) -> str:
        while True:
            line_end = buffer.find("\r\n")
            if line_end < 0:
                return buffer
            line = buffer[:line_end]
            upper = line.upper()
            if upper.startswith("MSG "):
                parts = line.split()
                size = int(parts[-1]) if parts and parts[-1].isdigit() else -1
                payload_start = line_end + 2
                payload_end = payload_start + size
                if size < 0 or len(buffer) < payload_end + 2:
                    return buffer
                payload = buffer[payload_start:payload_end]
                buffer = buffer[payload_end + 2 :]
                self._handle_payload(payload)
                continue
            buffer = buffer[line_end + 2 :]
            if upper == "PING":
                self._write_protocol("PONG\r\n")
            elif upper.startswith("-ERR"):
                self._emit_close(RuntimeError(f"NATS error: {line}"))

    def _handle_payload(self, payload: str) -> None:
        try:
            parsed = json.loads(payload)
        except Exception:
            return
        if isinstance(parsed, dict) and parsed.get("type") == "modcdp.nats.hello":
            self.peer_seen.set()
            return
        message = parsed.get("message") if isinstance(parsed, dict) and parsed.get("type") == "modcdp.nats.message" else parsed
        if isinstance(message, dict):
            self._emit_recv(message)


def _connect_options() -> dict[str, Any]:
    return {"verbose": False, "pedantic": False, "lang": "modcdp", "version": "1", "protocol": 1}


def _normalize_nats_url(url: str, subject_prefix: str | None = None) -> tuple[str, str]:
    parsed = urlparse(url)
    query = parse_qs(parsed.query)
    subject = subject_prefix or (query.get("subject") or query.get("subject_prefix") or [None])[0]
    query.pop("subject", None)
    query.pop("subject_prefix", None)
    normalized_query = urlencode(query, doseq=True)
    normalized_path = parsed.path or ("/" if parsed.scheme in ("ws", "wss") else "")
    normalized_url = urlunparse((parsed.scheme, parsed.netloc, normalized_path, parsed.params, normalized_query, parsed.fragment))
    return normalized_url, _sanitize_subject_prefix(subject or DEFAULT_NATS_SUBJECT_PREFIX)


def _sanitize_subject_prefix(value: str) -> str:
    subject = value.strip()
    if not subject or any(char.isspace() or char in "*>" for char in subject):
        raise ValueError(f"Invalid NATS subject prefix {value}")
    return subject
